### PR TITLE
Implemented the regex sub dialog under Prepare > Column:Text > Find/Replace Dialogue

### PR DESCRIPTION
**Pull request by N-thony**
Fixes (partially) #6893
@lloyddewit  @rdstern  this PR replaces PR #6994 due to the issue coming from 66 files as explained in that PR.
@rdstern you can still test this. But they are still changes you raised on items b) c) and d) I need to resolve after a quick discussion for clarification. Thanks.

### CODE REVIEW

**Comment by N-thony** [comment]
@rdstern as we discussed I added the regex function when it is chosen and this resolves your comment b) in PR #6994, please take a look.
I'm not yet sure on the position of the Boundary, perhaps we can also discuss about it and how it works?

**Review by rdstern — CHANGES_REQUESTED** [review]
@N-thony they all seem to work.  Well done.
I am now querying "the icing on the cake".  I think you have to give coll when using our default, because the stringr default is regex.  If easy, when it uses the default, could the command be given more simply?  In particular I think that when using regex and case = FALSE, then the command could be very simple?
Then happy when you can add the boundary.  I am not completely sure what it does - perhaps you are?

**Comment by N-thony** [comment]
> 

@lloyddewit I need to make the text change of `Word Count` to `Count` as @rdstern suggested during our discussion, I did the text change but it still shows `World Count` and this comes from the translation. As we did recently, if you agree, this can be done in a separate PR by setting rdoCount `DoNotTranslate`?

**Comment by rdstern** [comment]
@lloyddewit but here we would like a translation and I think this may already be "in the system" and not implemented yet.  At least it was requested earlier.  It is an example where we do (I think) still want to translate, but the changed text.  Currently it is Word Count - translated as Nombre de mots.  I suggest just Count, because it could be counting the number of characters, or words or sentences.  I assume that is translated into French as Nombre instead.

**Comment by N-thony** [comment]
> @lloyddewit but here we would like a translation and I think this may already be "in the system" and not implemented yet. At least it was requested earlier. It is an example where we do (I think) still want to translate, but the changed text. Currently it is Word Count - translated as Nombre de mots. I suggest just Count, because it could be counting the number of characters, or words or sentences. I assume that is translated into French as Nombre instead.

And if @lloyddewit agree to change it manually, the correct translation in French I think it will be `Compter` rather than `Nombre` according to what it is doing and making it consistent with other radio buttons text in French translated.

**Comment by N-thony** [comment]
> @N-thony they all seem to work. Well done. I am now querying "the icing on the cake". I think you have to give coll when using our default, because the stringr default is regex. If easy, when it uses the default, could the command be given more simply? In particular I think that when using regex and case = FALSE, then the command could be very simple? Then happy when you can add the boundary. I am not completely sure what it does - perhaps you are?

@rdstern please take a look at the output r code. For the implementation of the Boundary, more change in the design part, perhaps in another issue?

**Review by rdstern — CHANGES_REQUESTED** [review]
@N-thony I was trying to check the boundary control, but it wasn't enabled?
You also said that boundary was useful in the split dialogue, but I don't even see the regex there anymore.  Good to change the name of this PR?

**Comment by N-thony** [comment]
> @N-thony I was trying to check the boundary control, but it wasn't enabled? You also said that boundary was useful in the split dialogue, but I don't even see the regex there anymore. Good to change the name of this PR?

@rdstern you are right, I will add the boundary same time with regex in the Split dialogue after merging this if all is ok.

**Comment by rdstern** [comment]
ok. Can you make an issue with what remains on boundary and the split.  Then I will approve so the code can be checked too.

**Review by rdstern — DISMISSED** [review]
This is a very useful advance.  The boundary control is yet to be finalised, but that is being addressed in another issue.  So I am approving.  This is a good one to add!

**Comment by lloyddewit** [comment]
> > @lloyddewit but here we would like a translation and I think this may already be "in the system" and not implemented yet. At least it was requested earlier. It is an example where we do (I think) still want to translate, but the changed text. Currently it is Word Count - translated as Nombre de mots. I suggest just Count, because it could be counting the number of characters, or words or sentences. I assume that is translated into French as Nombre instead.
> 
> And if @lloyddewit agree to change it manually, the correct translation in French I think it will be `Compter` rather than `Nombre` according to what it is doing and making it consistent with other radio buttons text in French translated.

@N-thony Please could you:

- [ ] Change the translation in the translations database
- [ ] Change the translation in the CrowdIn system (this is needed, otherwise the database will revert to the old translation the next time we update the database from the CrowdIn JSON file)

Thanks

**Comment by lloyddewit** [comment]
@HawardKetoyoMsatsi Please could you peer review? thanks

**Comment by lloyddewit** [comment]
> Change the translation in the CrowdIn system (this is needed, otherwise the database will revert to the old translation the next time we update the database from the CrowdIn JSON file)

@N-thony The new English text is ' Count'. This does not yet exist CrowdIn so you'll need to take some extra steps:

- [ ] add ' Count' to `r_instat_not_menus.json'. When this is merged, it can be imported into CrowdIn. I think this happens automatically, but you may need to ask @dannyparsons . When it's in CrowdIn, you can change the French translation to ' Compter'.
- [ ] Add a new row for ' Count' to `translationsManuallyEntered.csv`


image

**Comment by HawardKetoyoMsatsi** [comment]
I Still find this issue. When I try searching for " double quotes using the escape character, after running when I reopen the dialogue doesn't remember double quotes.
What I run
image

When I reopen the dialog
image

Could this be an issue with string handling properties?

**Review by HawardKetoyoMsatsi — CHANGES_REQUESTED** [review]

File: instat/dlgStringHandling.vb [hunk 1]
```diff
@@ -124,16 +149,26 @@ Public Class dlgStringHandling
         ucrInputPattern.Reset()
         ucrReceiverForRegexExpression.ResetText()
 
+        clsDummyFunction.AddParameter("checked", False, iPosition:=0)
+
```
Comment (HawardKetoyoMsatsi) [inline instat/dlgStringHandling.vb:153]: About I don't know whether double double blank lines are encouraged in R-Instat if not then I guess you will have to remove one blank line here too.

**Review by HawardKetoyoMsatsi — CHANGES_REQUESTED** [review]

**Comment by N-thony** [comment]
@HawardKetoyoMsatsi thanks for your comments. I resolved them, please take a look.

**Comment by N-thony** [comment]
@HawardKetoyoMsatsi thanks for your comments. I resolved them, please take a look.

**Comment by HawardKetoyoMsatsi** [comment]
@N-thony great, it now works fine.
@lloyddewit over to you.

**Review by rdstern — DISMISSED** [review]
@N-thony I am happy to approve.  But you started with the Split dialogue and that now doesn't yet have the keyboard.  
Would you like to make that a separate issue, or would you prefer to add it now?

**Comment by N-thony** [comment]
> @N-thony I am happy to approve. But you started with the Split dialogue and that now doesn't yet have the keyboard. Would you like to make that a separate issue, or would you prefer to add it now?

Thanks, @rdstern, I would prefer now to add it then.

**Comment by lloyddewit** [comment]
I updated issue 6893 to show that this PR only partly solves it.

File: instat/sdgConstructRegexExpression.vb [hunk 2]
```diff
@@ -1,3 +1,105 @@
...
+' MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+' GNU General Public License for more details.
+'
+' You should have received a copy of the GNU General Public License 
+' along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
+Public Class sdgConstructRegexExpression
+
+    Private Sub sdgConstructRegexExpression_Load(sender As Object, e As EventArgs) Handles MyBase.Load
+        ttAddKeyboard.SetToolTip(cmdAny1, " Matches any character. For example,str_count(c(""rss"",""r.s.t""),""."") gives 3 5.  Using \\. Gives 0 2 ")
```
Comment (lloyddewit) [inline instat/sdgConstructRegexExpression.vb:20]: Extra spaces complicate the translations. Remove spaces at start and end?

File: instat/sdgConstructRegexExpression.vb [hunk 3]
```diff
@@ -1,3 +1,105 @@
...
+        ttAddKeyboard.SetToolTip(cmdbegin, " Start of the string. For example, str_count(c(""rss"",""r.s.t""),""^r\\."") gives 0 1")
```
Comment (lloyddewit) [inline instat/sdgConstructRegexExpression.vb:21]: remove space at start?

File: instat/sdgConstructRegexExpression.vb [hunk 4]
```diff
@@ -1,3 +1,105 @@
...
+        ttAddKeyboard.SetToolTip(cmdOr1, "or. For example, str_count(c(""rss"", ""rsstt""),""[st]"") gives 2 4")
+        ttAddKeyboard.SetToolTip(cmdNot1, "not any. For example, str_remove_all(c(""rs5"",""r19s20t""),""[^\\d]"") gives 5 1920")
+        ttAddKeyboard.SetToolTip(cmdOr3, "or. For example, str_count(c(""-abc"",""67""),""-|\\d"") gives 1 2")
+        ttAddKeyboard.SetToolTip(cmdOr2, "n times. For example, str_count(c(""bt"",""bat"",""boot"",""boat""),""b[ao]{1}t"") gives 0 1 0 0")
+        ttAddKeyboard.SetToolTip(cmdNumbers, "between. For example, str_count(c(""bt"",""bat"",""boot"",""boat""),""b[ao]{0,2}t"") gives 1 1 1 1")
+        ttAddKeyboard.SetToolTip(cmdZeroOrOne, "0 or 1 times. For example, str_count(c(""bt"",""bat"",""boot"",""boat""),""b[ao]?t"") gives 1 1 0 0 ")
+        ttAddKeyboard.SetToolTip(cmdPlusOne, "1 or more times. For example, str_count(c(""bt"",""bat"",""boot"",""boat""),""b[ao]+t"") gives 0 1 1 1")
+        ttAddKeyboard.SetToolTip(cmdZero, "0 or more times. For example, str_count(c(""bt"",""bat"",""boot"",""boat""),""b[ao]*t"") gives 1 1 1 1 ")
```
Comment (lloyddewit) [inline instat/sdgConstructRegexExpression.vb:32]: remove space at end?

File: instat/sdgConstructRegexExpression.vb [hunk 5]
```diff
@@ -1,3 +1,105 @@
...
+        ttAddKeyboard.SetToolTip(cmdDigit, "Digit (0 to 9), For example, str_replace_all(c(""rs5"",""r19s20t""),""[^\\d]"","") gives 5 1920")
+        ttAddKeyboard.SetToolTip(cmdSpace, "space. For example. str_remove_all(c(""r ss"",""r s t""),""\\s"") gives ""rss"" ""rst""")
+        ttAddKeyboard.SetToolTip(cmdOr1, "or. For example, str_count(c(""rss"", ""rsstt""),""[st]"") gives 2 4")
+        ttAddKeyboard.SetToolTip(cmdNot1, "not any. For example, str_remove_all(c(""rs5"",""r19s20t""),""[^\\d]"") gives 5 1920")
+        ttAddKeyboard.SetToolTip(cmdOr3, "or. For example, str_count(c(""-abc"",""67""),""-|\\d"") gives 1 2")
+        ttAddKeyboard.SetToolTip(cmdOr2, "n times. For example, str_count(c(""bt"",""bat"",""boot"",""boat""),""b[ao]{1}t"") gives 0 1 0 0")
+        ttAddKeyboard.SetToolTip(cmdNumbers, "between. For example, str_count(c(""bt"",""bat"",""boot"",""boat""),""b[ao]{0,2}t"") gives 1 1 1 1")
+        ttAddKeyboard.SetToolTip(cmdZeroOrOne, "0 or 1 times. For example, str_count(c(""bt"",""bat"",""boot"",""boat""),""b[ao]?t"") gives 1 1 0 0 ")
```
Comment (lloyddewit) [inline instat/sdgConstructRegexExpression.vb:30]: remove space at end?

File: instat/sdgConstructRegexExpression.vb [hunk 6]
```diff
@@ -1,3 +1,105 @@
...
+        ttAddKeyboard.SetToolTip(cmdPlusOne, "1 or more times. For example, str_count(c(""bt"",""bat"",""boot"",""boat""),""b[ao]+t"") gives 0 1 1 1")
+        ttAddKeyboard.SetToolTip(cmdZero, "0 or more times. For example, str_count(c(""bt"",""bat"",""boot"",""boat""),""b[ao]*t"") gives 1 1 1 1 ")
+        ttAddKeyboard.SetToolTip(cmdEscape, "Escape (with +*.? etc). For example, str_detect(c(""b$t"",""bat?"",""3*4""),""[\\?\\$\\*]"") gives TRUE TRUE TRUE")
+        ttAddKeyboard.SetToolTip(cmdPlusZero, "range of values. For example, str_count(c(""b$t"",""Bat?""),""[a-zA-Z]"") gives 2 3")
+
```
Comment (lloyddewit) [inline instat/sdgConstructRegexExpression.vb:35]:
```suggestion
```

**Review by lloyddewit — CHANGES_REQUESTED** [review]

**Review by lloyddewit — APPROVED** [review]

**Comment by N-thony** [comment]
@lloyddewit thanks for your review and the quick chat, I made the change and this is ready for @rdstern's review.

**Review by rdstern — APPROVED** [review]
Looks good now

**Comment by lloyddewit** [comment]
@N-thony This is tested and reviewed. Unfortunately, a merge conflict was triggered by another PR. Please could you resolve so that we can merge? Thanks

**Comment by N-thony** [comment]
> @N-thony This is tested and reviewed. Unfortunately, a merge conflict was triggered by another PR. Please could you resolve so that we can merge? Thanks

@lloyddewit I resolved the conflict. Thanks.

**Comment by lloyddewit** [comment]
@N-thony I still see 5 merge conflicts, maybe they're new? please could you fix? thanks

**Comment by N-thony** [comment]
> @N-thony I still see 5 merge conflicts, maybe they're new? please could you fix? thanks

Thanks, @lloyddewit, I resolved the conflicts. Please take a look.

**Review by lloyddewit — APPROVED** [review]

**Comment by lloyddewit** [comment]
@rdstern Since your last approval, there have been some merge conflicts resolved and other (hopefully non-functional) changes. Please can you test/approve? thanks

**Review by rdstern — CHANGES_REQUESTED** [review]
@N-thony I now get build errors

**Comment by N-thony** [comment]
> 

@rdstern it runs ok on my side. @lloyddewit could you try to run on your side. Thanks.

**Comment by N-thony** [comment]
> > 
> 
> @rdstern it runs ok on my side. @lloyddewit could you try to run on your side? Thanks.

@rdstern I will close this one and create another to avoid the build error.

**Comment by rdstern** [comment]
@N-thony thanks.  It will be good to get this merged.

**Comment by N-thony** [comment]
> > Change the translation in the CrowdIn system (this is needed, otherwise the database will revert to the old translation the next time we update the database from the CrowdIn JSON file)
> 
> @N-thony The new English text is ' Count'. This does not yet exist in CrowdIn so you'll need to take some extra steps:
> 
> * [ ]  add ' Count' to `r_instat_not_menus.json'. When this is merged, it can be imported into CrowdIn. I think this happens automatically, but you may need to ask @dannyparsons . When it's in CrowdIn, you can change the French translation to ' Compter'.
> * [ ]  Add a new row for ' Count' to `translationsManuallyEntered.csv`
> 
> image

@lloyddewit I think I have missed this part of adding `Count` to  `r_instat_not_menus.json`. Could this be done same time in PR #7108? Thanks.

**Comment by N-thony** [comment]
I have closed this one since it is replaced by #7118

**Pull request closed without merging**